src/server/database/database_provider.py

```diff
@@ -0,0 +1,114 @@
+            "SELECT name FROM person WHERE age=?", (age,)
+        ) as response:
+            if response.message is Message.OK:
+                name: list[str] = response.data
+            else:
+                # handle query failure
+    """
+
+    connection: sqlite3.Connection = sqlite3.Connection(":memory:")
+    database_source: str = ""
+
+    @classmethod
+    def initialize(cls) -> None:
+        """Initialize connection to the database."""
+        cls.database_source = PATHS.DATABASE
+
+        if not os.path.exists(cls.database_source):
+            result = cls._create_database()
+            if result is Message.NO_CONNECTION:
+                print("INFO: connecting to in-memory database as fallback")
+                cls.database_source = ":memory:"
+                cls._create_database()
+        cls._fill_database()
+
+    @classmethod
+    @contextmanager
+    def query(cls, query: str, params: tuple = ()) -> Generator[DatabaseResponse, None, None]:
+        """Interface that allows safe query execution in database.
+
+        Args:
+            query (str): Query to be executed in the database.
+            params (tuple, optional): Parameters passed to sqlite3 query string. Defaults to ().
+
+        Yields:
+            DatabaseResponse: Batch containing query execution status and received data (if any).
+
+        """
+        with cls._try_get_cursor() as cursor:
+            if cursor is None:
+                yield DatabaseResponse(Message.NO_CONNECTION, [])
+            else:
+                try:
+                    print(f"INFO: {query=}, {params=}")
+                    if params:
+                        cursor.execute(query, params)
+                    else:
+                        cursor.execute(query)
+                    data: list = cursor.fetchall()
+                    yield DatabaseResponse(Message.OK, data)
+                except sqlite3.IntegrityError as err:
+                    print(f"ERROR: {err}")
+                    yield DatabaseResponse(Message.INVALID_SQL, [])
+                except sqlite3.ProgrammingError as err:
+                    print(f"ERROR: {err}")
+                    yield DatabaseResponse(Message.INVALID_BINDINGS, [])
+                except sqlite3.Error as err:
+                    print(f"ERROR: {err}")
+                    yield DatabaseResponse(Message.UNKNOWN_ERROR, [])
+                finally:
+                    print(f"INFO: successfully executed a query: '{query}'")
+
+    @classmethod
+    @contextmanager
+    def _try_get_cursor(cls) -> Generator[Optional[sqlite3.Cursor], None, None]:
+        cursor: Optional[sqlite3.Cursor] = None
+        try:
+            cursor = cls.connection.cursor()
+            yield cursor
+        except sqlite3.Error as err:
+            print(f"ERROR: {err}")
+            yield None
+        finally:
+            if cursor is not None:
+                cursor.close()
+                cls.connection.commit()
```

Comment:
Do we have a use case where more than one query needs to be executed? If not, this type of interface would be enough.

In matter of transaction control, the purpose of DatabaseProvider is to hide away any SQL interface so that everything's taken care of under the hood.

On another note, we will probably replace `sqlite3` module with `psycopg2` as we need to move the database to the cloud in near future.

In case of error, `connection.commit()` does not do anything, as there is no transaction.

Reply:
Proponuję dodać parametr auto_commit do funkcji query, oraz do funkcji tworzącej kursor. Oraz dodatkowo funkcję commit.

Reply:
Frankly, I am not a fan of this solution because of the following reason: there is no way to enforce calling commit when auto_commit is set to false. This is a bug that virtually no change will occur in the database even though the query will have been executed by sqlite3 api.

Also if you want a fine-grained control of the transaction, just call `DatabaseProvider._try_get_cursor()` and you can execute as many queries as you want (with error handling!). Of course, calling private function is discouraged in comparison to the public interface of the class, but imo it is a good design.

Reply:
connection.commit() po prostu robi commit. Analogicznie można wykonać rollback, a tu tej możliwości nie mamy.
Wywoływanie metody prywatnej nie należy do dobrych (dlatego są one prywatnymi) - poza tym jest tam implikacja commit'a, a to nie zawsze jest to, czego się chce, szczególnie w przypadku błędu. 
Zawsze jak się tworzy klasy sterujące bazami danych, to tworzy się obsługę transakcji z wyższego poziomu.
Obecna implementacja zbytnio ukróca możliwości względem pojedynczej funkcji tworzącej połączenie z bazą.

Reply:
Do we have a use case for omitting commit and using rollback?

And as it was already said, in case of error, commit does not do anything.

Reply:
Aktualnie nie jest zaimplementowane całe api, więc mogą się pojawić potrzeby wykonywania rollback'u. 
Poza tym to nie jest kwestia przypadku użycia, a złego projektu interface'u.
No i się nie zaleca, aby commit był automatycznie implikowany. Dopiero kiedy wyższe api stwierdzi, że wszystko jest zrobione, powinien on być wywołany.

Chyba, że w przypadku wprowadzenia części danych i wyrzucenia błędu, sekcja finally jednak się nie wykona i nie będzie commitu.

Reply:
You are missing the point. Its purpose is not to have another abstraction layer which virtually changes nothing in case of the interface. It's about hiding any SQL handling inside this class, so that a user of this class does not have to worry about commiting, not commiting, rollbacking (of which, so far, there is no need to use).

And as it was already said, in case of error, commit does not do anything.

Reply:
W takim razie nie powinien być usuwany moduł do uzyskiwania połączenia do bazy danych - a właściwie powinien być tu wykorzystany. 
Lub co najmniej powinna istnieć metoda publiczna do uzyskania tego połączenia.

Reply:
Why is there supposed to be a way to connect to the database in the public interface if we are ever going to use only one database? I just don't see the point of this. This operation can be hidden away in initialize() method and no external user has to be concerned about this.

Reply:
Właśnie z tych powodów, co piszę - mamy porządnie ucięte funkcjonalności.

Reply:
The functionality remains the same, it's just abstracted away, as we are not planning to expand support to multiple database targets.

Reply:
Brak możliwości wykonania stricte: rollback, commit, execute_many
Dalej nie chce mi się wymieniać

Reply:
Can you provide use cases in our code where these functionalities are required?

Reply:
Tak, przy tworzeniu wielu insertów, zalecane jest użycie executemany, zamiast jednego wielkiego zapytania insert.
Taki przypadek użycie prawdopodobnie będzie zachodził w api aktualizacyjnym kurs w bazie, oraz przewidującym kurs.
Także jakiekolwiek przypadku dwóch poleceń insert, muszą być one zawarte w jednej transakcji. Ewentualnie przy dwóch update-ach. Przypadek użycia? Zakup kryptowaluty - jednocześnie trzeba odjąć sumę w USD i dodać w BTC.

Reply:
I've changed the interface to fulfill the needs you listed. From now on, we have a direct access to `sqlite3.Cursor` instance. Error handling still works under the hood. The commit still happens when leaving context, but now we can commit any changes faster (if there's a need for that) and roll out multiple executions.